tooling/image-sync/repository.go

```diff
@@ -0,0 +1,363 @@
+	}
+
+	return &authSecret, nil
+}
+
+// EnsureRepositoryExists ensures that the repository exists
+func (a *AzureContainerRegistry) RepositoryExists(ctx context.Context, repository string) (bool, error) {
+
+	pager := a.acrClient.NewListRepositoriesPager(nil)
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return false, fmt.Errorf("failed to advance page: %v", err)
+		}
+		for _, v := range page.Repositories.Names {
+			if *v == repository {
+				return true, nil
+			}
+		}
+	}
+
+	return false, nil
+}
+
+func ptr[T any](v T) *T {
+	return &v
+}
+
+// GetTags returns the tags in the given repository
+func (a *AzureContainerRegistry) GetTags(ctx context.Context, repository string) ([]string, error) {
+
+	var tags []string
+
+	pager := a.acrClient.NewListTagsPager(repository, &azcontainerregistry.ClientListTagsOptions{OrderBy: ptr(azcontainerregistry.ArtifactTagOrderByLastUpdatedOnDescending)})
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to advance page: %v", err)
+		}
+		for _, v := range page.Tags {
+			if *v.Name == "latest" {
+				continue
+			}
+			tags = append(tags, *v.Name)
+		}
+		if len(tags) >= a.numberOfTags {
+			break
+		}
+	}
+
+	return tags, nil
+}
+
+// OCIRegistry implements OCI Repository access
+type OCIRegistry struct {
+	httpclient   *http.Client
+	baseURL      string
+	numberOftags int
+}
+
+// NewOCIRegistry creates a new OCIRegistry access client
+func NewOCIRegistry(cfg *SyncConfig, baseURL string) *OCIRegistry {
+	o := &OCIRegistry{
+		httpclient:   &http.Client{Timeout: time.Duration(cfg.RequestTimeout) * time.Second},
+		numberOftags: cfg.NumberOfTags,
+	}
+	if !strings.HasPrefix(o.baseURL, "https://") {
+		o.baseURL = fmt.Sprintf("https://%s", baseURL)
+	} else {
+		o.baseURL = baseURL
+	}
+	return o
+}
+
+type rawManifest struct {
+	TimeUploadedMs string
+	Tag            []string
+}
+
+type rawOCIResponse struct {
+	Manifest map[string]rawManifest
+	Tags     []string
+}
+
+func getNewestTags(response *rawOCIResponse, numberOfTags int) ([]string, error) {
+	var returnTags []string
+
+	uploadedTagAt := make(map[int][]string)
+	uploadTimes := make([]int, 0, len(response.Manifest))
+
+	for _, manifest := range response.Manifest {
+		if len(manifest.Tag) == 0 {
+			continue
+		}
+		uploadedAt, err := strconv.Atoi(manifest.TimeUploadedMs)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse manifest %s time: %v", manifest, err)
+		}
+		uploadedTagAt[uploadedAt] = manifest.Tag
+		uploadTimes = append(uploadTimes, uploadedAt)
+	}
+	sort.Sort(sort.Reverse(sort.IntSlice(uploadTimes)))
+
+	for i, k := range uploadTimes {
+		if i >= numberOfTags {
+			break
+		}
+		returnTags = append(returnTags, uploadedTagAt[k]...)
+	}
+
+	return returnTags, nil
+}
+
+// GetTags returns the tags in the given repository
+func (o *OCIRegistry) GetTags(ctx context.Context, image string) ([]string, error) {
```

Comment:
any reason to now use `docker.GetRepositoryTags()` from github.com/containers/image as a basis for the implementation?

Reply:
I considered it, but it returns a flat list of tags ([]string), ref: https://pkg.go.dev/github.com/containers/image/v5/docker#GetRepositoryTags

The basis for this sync however is to only sync N tags, ordered by update/creation. Thus, we need our own implementation here.